src/openfl/utils/_internal/ShaderMacro.hx

```diff
@@ -199,6 +442,211 @@ class ShaderMacro
 		return fields;
 	}
 
+	private static inline function getDefaultGLVersion():String
+	{
+		// Specify the default glVersion.
+		// We can use compile defines to guess the value that prevents crashes in the majority of cases.
+		return #if (android) "100" #elseif (web) "100" #elseif (mac) "120" #else "100" #end;
+	}
+
+	/**
+	 * Attempt to migrate GLSL code from the current (old) GLSL shader format to the specified (newer) one.
+	 * @param source The source to convert.
+	 * @param version The version to convert to.
+	 * @param isFragment Whether the source is a fragment shader. False if it is a vertex shader.
+	 * @return The converted source.
+	 */
+	private static function processGLSLText(source:String, glVersion:String, isFragment:Bool)
+	{
+		if (glVersion == "" || glVersion == null) return processGLSLText(source, getDefaultGLVersion(), isFragment);
+
+		// No processing needed on "compatibility" profile
+		if (StringTools.endsWith(glVersion, " compatibility")) return source;
+		if (StringTools.endsWith(glVersion, " core")) return processGLSLText(source, StringTools.replace(glVersion, " core", ""), isFragment);
+
+		// Recall: Attribute values are per-vertex, varying values are per-fragment
+		// Thus, an `out` value in the vertex shader is an `in` value in the fragment shader
+		var attributeKeyword:EReg = ~/attribute ([A-Za-z0-9]+) ([A-Za-z0-9_]+)/g; // g to match all
+		var varyingKeyword:EReg = ~/varying ([A-Za-z0-9]+) ([A-Za-z0-9_]+)/g; // g to match all
+
+		var texture2DKeyword:EReg = ~/texture2D/g;
+		var glFragColorKeyword:EReg = ~/gl_FragColor/g;
+
+		switch (glVersion)
+		{
+			default:
+				// We don't know this GLSL version, so don't do anything.
+				return source;
```

Comment:
Maybe show an explicit warning like "OpenFl doesn't know glsl version ${glVersion}, no conversion will happen" ?

Reply:
So then how do we use a certain glsl version in the project.xml or could it only be used in .hx code?

Reply:
You don't specify GLSL version in the project.xml you specify it when you construct the shader

Reply:
oh ok so then the default whould be 120 for desktops and for web it whould be 300?

Reply:
> oh ok so then the default whould be 120 for desktops and for web it whould be 300?

The default would be the result of `getDefaultGLSLVersion`

Reply:
ohhh ok

Reply:
then how whould i use a certain glsl version?

Reply:
> then how whould i use a certain glsl version?

You can use `@:glVersion("120")` the same way you use `@:glFragmentBody("...")`.

As a note there is an `FlxRuntimeShader` in `flixel-addons`. It was originally built by me and ran on this, but it was modified to not use this code so it could be merged. When this is merged it will get changed back and then it will support setting the `glVersion` among other things.